Update python versions [BF-2100]

Mulugruntz opened the pull request:
Python versions are now in sync between the `README`, the CI and the classifiers.

New:
- `setup.py` now has 3.9+.

To be noted:
- Python 3.11 binaries are available since `psycopg2>=2.9.5`. The version pinning is only for the CI. Setuptools still picks the latest compatible version for the platform (unchanged).

[BF-2100]

### Proposed changes in this pull request

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [x] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [x] Documentation Improvement
- [x] Other

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [x] All the tests are passing after the introduction of new changes.
- [x] Added tests respective to the part of code I have written.
- [x] Added proper documentation where ever applicable (in code and README.md).

### Optional extra information



[BF-2100]: https://aiven.atlassian.net/browse/BF-2100?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Update requirements versions [BF-2100]

Python 3.11 binaries are available since psycopg2>=2.9.5.

[BF-2100]

Mulugruntz commented:
@jlprat For this one, we'll need to modify the settings for the Required checks. This is because `lint` and `unit-test` won't exist anymore. They now have a matrix of python 3.9-11.

jlprat commented:
@Mulugruntz It makes sense :)

alexole commented on Makefile:
```diff
@@ -49,6 +49,9 @@ validate-style:
 	-rm $(CHANGES_BEFORE) $(CHANGES_AFTER)
 
+validate-python-version:
+	$(PYTHON) -c 'import sys; v=sys.version_info; assert v >= (3, 9), f"Python 3.9 or newer is required. Current version: {v[0]}.{v[1]}.{v[2]}."'
```
Will it really break if the version is lower? I don't think it's a good idea to hard fail if it most probably works but we don't test it.

Mulugruntz replied:
That's the idea, indeed. When it fails, it `exit(1)`.

Now in this PR, we do test in the CI.

About the "dropping versions", we were saying "3.5+" but I know for sure it was already not 3.5 compatible. For example, in `conftest.py`, we do `from __future__ import annotations`, which is 3.7+.

So, we're not necessarily dropping versions. Now, we have a real support scheme, and I think that's a better thing fur users (more reliable).

alexole replied:
I would not like it as a user, if I use old python version and it works even without guarantees I prefer it that way in some cases instead of artificial failing with unsupported version, when we know it will most probably work.There is no reason to hard fail. We can document what we support, but I don't think we should have this check unless it's really needed and is justifiable.

Mulugruntz replied:
This prevents us from evolving a pre-alpha 0.0.6 product.
If we do evolve it without "artificially breaking", that means we'll break it later without any warning (as these are untested versions).

As a user, I'd rather have something that I can rely on. If I'm being told that 0.0.7 is CPython 3.9+ only, then I can still stay on 0.0.6 or fork it myself. But I won't have guarantees that it'll work anymore.

Last but not least, we declared it as a pre-alpha version, we're free to break compatibility anytime (and that's what was confirmed in yesterday's Daily).

As a conclusion, between "we don't have any clear support scheme, try it and it might work this time but maybe not next update" and "okay, it's tested and supported on versions x, y, and z", I'd always put my money on the latter one, as a user.

Also: old EOL'd Python versions do not receive security updates. They're unsafe and should be avoided.

Mulugruntz replied:
@alexole I don't think it's worth debating over, for now. I removed the mandatory checks, just like you requested. And I left the target, so that users can still run it if they want. I hope this compromise is acceptable.

Also, I split the commits in a better way, and removed some deprecated parameters (for isort) that were spewing warnings. Hopefully, we can merge this before the end of the day ;-).

alexole replied:
Why do you need to keep this target here still? Supported version is mentioned in `setup.py`

Commit:
Update GitHub Action versions [BF-2100]

[BF-2100]

Commit:
Update Python versions [BF-2100]

Python versions are now in sync between the `README`, the CI and the classifiers.

New:
- `setup.py` now has 3.9+.

[BF-2100]

Mulugruntz commented:
Hi @jlprat , now that it's merged, feel free to update the repo settings to set the new checks as required.

jlprat commented:
@Mulugruntz these change can be done via PR. You can adapt the files under `.github/workflows`. In here you can add/remove/condition the current flows. No special permissions needed to do this change :)

Mulugruntz commented:
@jlprat I think I'm talking about something else.
If you go to the repo settings, in Branches, you'll see something like this:
<img width="1150" alt="image" src="https://github.com/aiven/aiven-db-migrate/assets/13810199/09bc0687-7398-41d6-98d9-57aeef853eb2">
The names of the status checks need to be updated. Because the matrix has changed. This cannot be changed in the `.github/workflows` afaik. Otherwise, it'll wait for some non-existent checks to finish... forever.
Some documentation [here](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches#require-status-checks-before-merging) and [there](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/troubleshooting-required-status-checks).

And a tutorial: https://betterprogramming.pub/github-status-checks-and-branch-protection-made-easy-b70d6d9ffc76
